### DISCUSSION TITLE
Labo 2 - rentFilm

### DISCUSSION THREAD

**Discussion by bchollet** [amt-classroom/amt-classroom.github.io#12] [Q&A]
Bonjour,

Nous n'arrivons pas à faire fonctionner la méthode rentFilm au moyen seul des transactions. Quoique nous faisions, nous avons toujours des écritures concurrentes qui ont été réalisée.

@bchapuis nous a donné la solution suivante qui ne lui convenait guère car "trop brutale": 
```java
em.createNativeQuery("LOCK TABLE rental IN EXCLUSIVE MODE").executeUpdate();
```

Y a-t-il une solution plus "élégante" ?

**Comment by Perdjesk** [comment]
Le test fournit n'a pas été actualisé pour la version solution qui utilise la sérialisation des transactions dans Quarkus.

Voici le patch à appliqué dans le test:

```java
@@ -67,14 +64,24 @@ class RentalServiceTest {
       var termination = executor.awaitTermination(30, TimeUnit.SECONDS);
       Assertions.assertTrue(termination);

+      AtomicInteger nbFailedInProxyTransactional = new AtomicInteger();
+
       var successfulRentals = futures.stream().filter( f -> {
         try {
           return f.get();
         } catch (Exception e) {
-          throw new RuntimeException(e);
+          // ignore
+          // Could fail due to pgsql serializable isolation level and predicate locking
+          // This exception happens in the proxy of @Transactional and can not be caught by the implementation method
+          // See: https://www.postgresql.org/docs/16/transaction-iso.html#XACT-SERIALIZABLE
+          nbFailedInProxyTransactional.getAndIncrement();
+          return false;
         }
       }).toList();

+      // Just to get a bit of data about how many failed through predicate locking.
+      System.out.println("----- nbFailedInProxyTransactional: "+ nbFailedInProxyTransactional);
+
       // Validate that only one rental for the same item was inserted in database

```

**Comment by Greywind299792458** [comment]
Bonjour @Perdjesk 
nous venons d'appliquer le fix dans notre repo et le test `testRental()` n'est toujours pas décidé à passer chez nous...

L'assertion qui échoue est la suivante : 
`Assertions.assertEquals(1, rentalTestHelper.getRentalsForInventory(inventoryId).size(), "DB, Only one rental active per inventory item (Pass:"+p+")");`

Nous avons toujours des valeurs supérieures à 1 ici.

Est-ce que nous n'avons pas bien géré au niveau de notre méthode de location dans notre service ou est-ce un problème ailleurs ?

Nous avons, en plus de l'annotation @Transactionnal, tenté d'utiliser la notion de `setLockMode`([doc utilisée](https://www.baeldung.com/jpa-pessimistic-locking)) au cas où l'annotation n'était pas suffisante ?? Mais sans succès jusqu'à présent...

Cordialement,

Elliot Ganty & Alexis Monthoux

**Reply by Perdjesk** [reply]
Le patch précèdent est uniquement nécessaire pour un type de solution spécifique.
Ce patch n'est pas en lien avec la solution donnée via l'utilisation d'un verrou exclusif sur la table.

Les deux aspects se sont mélangés dans cette question, je me permets d'apporter des détails.

Plusieurs solutions sont possibles afin de résoudre la concurrence de plusieurs transactions sur la même pièce d'inventaire, trois solutions différentes ont été validées lors de la réalisation du laboratoire.
La définition seule d'une transaction avec le niveau d'isolation par défaut, ne permet pas de garantir la résolution des accès concurrent dans ce cas.

@bchapuis a donné en cours la solution via le verrou exclusif sur la table entière. 

Une autre solution requiert le patch donné afin que le test soit encore valide. Les commentaires inclus dans le patch mettent sur la piste de cette solution.
La troisième solution qui a été validée, qui pourrait être jugée plus élégante que les deux dernières, emploi une approche différente aux verrous et au niveau d'isolation.

Durant la réalisation du laboratoire une solution via JPA `setLockMode` n'a pas permis d'implémenter une solution valide.
Voici la version qui a été explorée:
```java
              """, Inventory.class)
          .setParameter("inventoryId", inventory.getId())
          // The following won't work to prevent duplicates rentals of same inventory item.
          // TODO: develop why
          // .setLockMode(LockModeType.PESSIMISTIC_WRITE)
          .setMaxResults(1).getSingleResult();
```

La contrainte de temps n'a pas permis d'aboutir à une solution via cette approche et certaines limitations ont été identifiées à cette approche à cause de la nature du besoin pour cette transaction et le format des données dans Sakila qui n'a pas permis de bénéficier de cette fonctionnalité. Il semblerait qu'il ne soit pas réalisable de résoudre le besoin de cette requête au travers de cette fonctionnalité seule.
Si vous le souhaitez votre analyse et explication de pourquoi cette approche ne fonctionne pas comme espéré sont les bienvenues dans votre rapport et seront considérées pour l'évaluation de l'exercice 1.
Toutefois il est requis de proposé une version valide.
